### PR TITLE
Rust module API rework

### PR DESCRIPTION
**Pull request by coolreader18**
WIP - waiting on abi rework implementation.

### CODE REVIEW

File: crates/bindings/src/lib.rs
Comment (gefjon) [inline crates/bindings/src/lib.rs:100]: `DbContext` in modules also has methods `rng`, `identity` and `address`. `identity` and `address` are subject to change and need not be implemented yet, but `rng` is important and should be included.

File: crates/bindings/src/lib.rs
Comment (gefjon) [inline crates/bindings/src/lib.rs:285]: Now I am even more confused.

File: crates/bindings/src/lib.rs
Comment (gefjon) [inline crates/bindings/src/lib.rs:290]: This could potentially contain `table: String, column: String` for example. That can be future work, but please leave a TODO comment in here to get better error messages.

File: crates/bindings/src/lib.rs
Comment (gefjon) [inline crates/bindings/src/lib.rs:383]: @Centril will this still be true with the new ABI?

File: crates/bindings/src/lib.rs
Comment (gefjon) [inline crates/bindings/src/lib.rs:389]: Impl this in terms of `self.delete` and an insert. There's not going to be a host call which implements this method.

File: crates/bindings/src/lib.rs
Comment (gefjon) [inline crates/bindings/src/lib.rs:408]:
```suggestion
            todo!("Implement this method once the new ABI exposes physical index ops to WASM");
```

File: crates/bindings/src/lib.rs
Comment (gefjon) [inline crates/bindings/src/lib.rs:413]:
```suggestion
            todo!("Implement this method once the new ABI exposes physical index ops to WASM");
```

File: crates/bindings/src/rt.rs
Comment (gefjon) [inline crates/bindings/src/rt.rs:220]: I believe this branch is no longer valid because the `ctx` arg is now mandatory.

File: crates/bindings/src/table.rs
Comment (gefjon) [inline crates/bindings/src/table.rs:70]: Again, not how the `Table` trait is specified.

File: crates/bindings/src/table.rs
Comment (gefjon) [inline crates/bindings/src/table.rs:280]: Neat macro-ology!

File: crates/bindings/src/table.rs
Comment (gefjon) [inline crates/bindings/src/table.rs:298]: We need to discuss (on the proposal) whether we actually want to accept all combinations of `T` and `&T`.

**Review by gefjon — CHANGES_REQUESTED** [review]
- You still have the macros looking for known concrete types for columns marked `autoinc`.
- I would really appreciate some comments and docs on the macros. In particular, listing some examples of expected inputs and outputs, and some explanation of why the different parts of the outputs are there and what parts of the runtime system they plug into.

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs:661]: I had hoped/thought so, but in practice, no, you have to import the trait to call the methods.

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs:651]: Both. I'll change it to require it be specified.

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs:137]: Both. I was trying to change the macros as little as possible, since there's already #1314 which I had hoped would be able to land first.

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs:662]: Annoying hacky way to get around a compilation error until rust-lang/rfcs#3498 drops

File: crates/bindings/src/rt.rs
Comment (coolreader18) [reply crates/bindings/src/rt.rs:220]: Ah, I missed that

File: crates/bindings/src/table.rs
Comment (coolreader18) [reply crates/bindings/src/table.rs:70]: I couldn't tell how it was supposed to be from the proposal - "For tables with constraints, [TryInsertError] is an enum generated by the `#[spacetimedb::table]` macro. [...] TODO: specify a generated name for this type?". In the interest of generating as few identifiers as possible, I made it so that it would be an enum defined in the bindings crate, but with variants defined based on the given `Table`. Are you amenable to using this implementation? IMO it's a better UX & better docs than the generated enum version mentioned in the proposal.

File: crates/bindings/src/lib.rs
Comment (gefjon) [reply crates/bindings/src/lib.rs:100]: If we add the trait later and move inherent methods to trait methods, that's a breaking change.

File: crates/bindings/src/table.rs
Comment (gefjon) [reply crates/bindings/src/table.rs:70]: I am amenable to this. Leave a comment on the proposal and I'll update it to reflect your version.

File: crates/bindings-macro/src/lib.rs
Comment (gefjon) [reply crates/bindings-macro/src/lib.rs:661]: Do you think using an `impl Trait` type here is tangibly better in some way than a concrete type? If so, I'm open to revising the proposal, but my intuition is that we should just return a `BTreeIndex<#row_type, #index_ty>`.

File: crates/bindings-macro/src/lib.rs
Comment (gefjon) [reply crates/bindings-macro/src/lib.rs:662]: Can you elaborate? What's the compiler error, and how does adding this do-nothing trait fix it?

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs:661]: Hm, I guess that's good - I assumed the proposal meant an autogenerated concrete type, but if it can be a lib-defined concrete type, then that sounds good.

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs:662]: It's explained more in depth in that [rfc](https://rust-lang.github.io/rfcs/3498-lifetime-capture-rules-2024.html), but basically - return types often "capture" lifetime & type parameters, which for a concrete type means like, mentioning `'a` in the type (`fn foo() -> X<'a>`). impl traits represent opaque types, which are implicitly defined and thus need to have rules of how and when to capture lifetimes. If the opaque type doesn't capture the lifetime, but the underlying type it represents does, that's a compile error. In edition2024, those rules will be changed to capture all in-scope lifetimes by default, with the opt-in ability to specify a specific set of lifetimes to capture instead (`fn foo() -> impl Trait + use<'a>`). For now, we have to polyfill that behavior by adding a dummy trait that just mentions the lifetime, but does nothing else - telling the compiler that the opaque type captures the lifetime.

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs:661]: The main thing is that we have to expose the implementation detail of the `const COL_IDX: u16` arg, but I think that's fine.

File: crates/bindings-macro/src/lib.rs
Comment (gefjon) [reply crates/bindings-macro/src/lib.rs:662]: I thought that `impl Trait + 'a` already did this.

File: crates/bindings/src/lib.rs
Comment (coolreader18) [reply crates/bindings/src/lib.rs]: Well, it wouldn't necessarily be moving them - we could just keep the inherent methods and add the trait impl. From a user perspective, having to always import this weird `DbContext` trait just to use an rng seems unergonomic. I think using trait syntax to define this API is really helpful, but that doesn't mean we have to have a singleton trait with only one impl in the actual crate.

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs:661]: oh, that's why I assumed :upside_down_face:

> ```rust
> // Name may vary. Type may be generic over a lifetime bound.
> struct MyTableMyUniqueColumnUniqueIndex {
>     // private fields
> }
> 
> impl TableHandle {
>     // May return a shared reference `&MyTableMyUniqueColumnUniqueIndex`.
>     fn my_unique_column(&self) -> MyTableMyUniqueColumnUniqueIndex;
> }
> ```

Comment (gefjon) [reply crates/bindings-macro/src/lib.rs:661]: Ah, this is more than I intended to specify. Returning a concrete library type is fine. I'll make a note to update the proposal.

File: crates/bindings/src/lib.rs
Comment (gefjon) [reply crates/bindings/src/lib.rs]: Does the same reasoning apply to clients, where you have to import the `DbContext` trait in order to do things like subscribe to queries and inspect the connection's status?

File: crates/bindings-macro/src/lib.rs
Comment (coolreader18) [reply crates/bindings-macro/src/lib.rs]: Welp, they're gone anyway, since I switched over to concrete types.

**Comment by coolreader18** [comment]
Currently just blocked on #1704.

**Comment by coolreader18** [comment]
I guess now it's just blocked on an implementation in ModuleHost

**Comment by coolreader18** [comment]
Differences from proposal:
- TryInsertError is a generic enum, not an associated type (this has since been amended in the proposal)
  - `Table::UniqueConstraintViolation` and `AutoIncOverflow` associated types that define each variant of the TryInsertError enum.
  - `ctx.db.my_table()` returns a `&'_ my_table_Handle`, not a `my_table_Handle<'_>` (this is so that `TryInsertError<Self>` is trivially `'static`, and can be returned as a `Box<dyn Error>`)

**Comment by coolreader18** [comment]
The merge is just for testing - I'll rebase once #1697 actually merges.

**Comment by coolreader18** [comment]
@Centril do you know why this might be failing? In `BTreeIndex::filter()`:

```rust
        let index_id = Idx::index_id();
        let args = b.get_args();
        let (prefix, prefix_elems, rstart, rend) = args.args_for_syscall();
        let iter = sys::datastore_btree_scan_bsatn(index_id, prefix, prefix_elems, rstart, rend)
            .unwrap_or_else(|e| panic!("unexpected error from datastore_btree_scan_bsatn: {e}"));
        TableIter::new(iter)
```
It panics not at the lookup of the index id by name (searches the system tables), but in datastore_btree_scan_bsatn with "No such index" (searches `index_id_map`). Is the index not getting properly committed to the database state? As far as I can tell it is, but I don't know why else this would be failing.

**Comment by coolreader18** [comment]
gruhhh.... the tests are failing because the rust module now declares snake_case table names but the csharp module doesn't

**Comment by coolreader18** [comment]
theoretically this is all tests passing, besides that

**Comment by coolreader18** [comment]
maybe

**Comment by coolreader18** [comment]
yeah, rust tests are all passing for me locally

**Comment by coolreader18** [comment]
Reviewer's discretion whether this is an acceptable solution, I guess.

**Comment by gefjon** [comment]
Can you alter the C# module to define snake_case table names by changing `class MyTable` to `class my_table`?

**Review by gefjon — APPROVED** [review]

**Comment by gefjon** [comment]
Don't merge until we have a companion PR for private.

**Pull request merged**
